jsonschema-cli: init at 0.29.0

kachick opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR aims to introduce https://github.com/Stranger6667/jsonschema/tree/rust-v0.29.0/crates/jsonschema-cli as a new package.

```
A fast command-line tool for JSON Schema validation, powered by the jsonschema crate.
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kachick committed:
jsonschema-cli: init at 0.29.0

kachick commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382774`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>jsonschema-cli</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>jsonschema-cli</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>jsonschema-cli</li>
  </ul>
</details>

Tert0 commented on pkgs/by-name/js/jsonschema-cli/package.nix:
```diff
@@ -0,0 +1,66 @@
+  pname = "jsonschema-cli";
+  version = "0.29.0";
+
+  src = fetchCrate {
+    inherit pname version;
+    hash = "sha256-kpdvvCnMsHfogXmAqNeo1Cl1hZtCPHqkfhYm8ipWToo=";
+  };
```
Consider fetching the source from GitHub. Afaik using Github is prefered

kachick replied:
I have preferred using crate because of currently the GitHub repository does not save the Cargo.lock. Are there any documents explaining how to determine these methods?

Tert0 replied:
Ok, this is a good reason to use the crate

Tert0 commented on pkgs/by-name/js/jsonschema-cli/package.nix:
```diff
@@ -0,0 +1,66 @@
+    versionCheckHook
+  ];
+  doInstallCheck = true;
+  versionCheckProgram = "${placeholder "out"}/bin/${meta.mainProgram}";
```
I think it is discouraged to use mainProgram here. The version check works without this line.

kachick replied:
https://github.com/NixOS/nixpkgs/blob/6d0e5aea88d0f8fa0d0e1258c65c4f799c46717e/doc/hooks/versionCheckHook.section.md?plain=1#L34


The default depends on `pname`. It sounds robust to me to specify `meta.mainProgram`.

SuperSandro2000 commented on pkgs/by-name/js/jsonschema-cli/package.nix:
```diff
@@ -0,0 +1,66 @@
+  versionCheckProgramArg = [ "--version" ];
+
+  passthru = {
+    tests = {
```
Are there no upstream tests?

kachick replied:
Looks having snapshot tests https://github.com/Stranger6667/jsonschema/tree/rust-v0.29.0/crates/jsonschema-cli/tests for more matrix. I have added these two tests because I want to make sure the basic usage and behaviors has not been changed in PR, especially by the r-ryantm bot.

SuperSandro2000 replied:
So we are not running the upstream tests? otherwise it is pretty pointless to add our own bare bones tests.

kachick replied:
Sorry if I misunderstood. I expect this package definition includes and runs the upstream snapshot tests.

From the nixpkgs-review log.

```
Using versionCheckHook
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/7v55q4djq8l6sxdpjgqj7qkrwpvb8mb4-jsonschema-cli-0.29.0.tar.gz
source root is jsonschema-cli-0.29.0.tar.gz
Executing cargoSetupPostUnpackHook
Finished cargoSetupPostUnpackHook
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Executing cargoSetupPostPatchHook
Validating consistency between /build/jsonschema-cli-0.29.0.tar.gz/Cargo.lock and /build/jsonschema-cli-0.29.0-vendor/Cargo.lock
Finished cargoSetupPostPatchHook
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing cargoBuildHook
cargoBuildHook flags: -j 4 --target x86_64-unknown-linux-gnu --offline --profile release
[1m[32m   Compiling[0m proc-macro2 v1.0.93
[1m[32m   Compiling[0m unicode-ident v1.0.16
[1m[32m   Compiling[0m autocfg v1.4.0
[1m[32m   Compiling[0m libc v0.2.169
[1m[32m   Compiling[0m stable_deref_trait v1.2.0
[1m[32m   Compiling[0m memchr v2.7.4
[1m[32m   Compiling[0m smallvec v1.13.2
[1m[32m   Compiling[0m num-traits v0.2.19
[1m[32m   Compiling[0m quote v1.0.38
[1m[32m   Compiling[0m syn v2.0.98
[1m[32m   Compiling[0m serde v1.0.217
[1m[32m   Compiling[0m writeable v0.5.5
[1m[32m   Compiling[0m itoa v1.0.14
[1m[32m   Compiling[0m litemap v0.7.4
[1m[32m   Compiling[0m slab v0.4.9
[1m[32m   Compiling[0m pin-project-lite v0.2.16
[1m[32m   Compiling[0m futures-core v0.3.31
[1m[32m   Compiling[0m icu_locid_transform_data v1.5.0
[1m[32m   Compiling[0m futures-sink v0.3.31
[1m[32m   Compiling[0m futures-io v0.3.31
[1m[32m   Compiling[0m icu_properties_data v1.5.0
[1m[32m   Compiling[0m pin-utils v0.1.0
[1m[32m   Compiling[0m futures-task v0.3.31
[1m[32m   Compiling[0m fnv v1.0.7
[1m[32m   Compiling[0m bytes v1.10.0
[1m[32m   Compiling[0m once_cell v1.20.3
[1m[32m   Compiling[0m futures-util v0.3.31
[1m[32m   Compiling[0m http v1.2.0
[1m[32m   Compiling[0m num-integer v0.1.46
[1m[32m   Compiling[0m socket2 v0.5.8
[1m[32m   Compiling[0m mio v1.0.3
[1m[32m   Compiling[0m synstructure v0.13.1
[1m[32m   Compiling[0m cfg-if v1.0.0
[1m[32m   Compiling[0m icu_normalizer_data v1.5.0
[1m[32m   Compiling[0m utf8_iter v1.0.4
[1m[32m   Compiling[0m write16 v1.0.0
[1m[32m   Compiling[0m percent-encoding v2.3.1
[1m[32m   Compiling[0m httparse v1.10.0
[1m[32m   Compiling[0m utf16_iter v1.0.5
[1m[32m   Compiling[0m tokio v1.43.0
[1m[32m   Compiling[0m http-body v1.0.1
[1m[32m   Compiling[0m lock_api v0.4.12
[1m[32m   Compiling[0m ryu v1.0.19
[1m[32m   Compiling[0m serde_json v1.0.138
[1m[32m   Compiling[0m try-lock v0.2.5
[1m[32m   Compiling[0m ref-cast v1.0.23
[1m[32m   Compiling[0m version_check v0.9.5
[1m[32m   Compiling[0m parking_lot_core v0.9.10
[1m[32m   Compiling[0m ahash v0.8.11
[1m[32m   Compiling[0m want v0.3.1
[1m[32m   Compiling[0m form_urlencoded v1.2.1
[1m[32m   Compiling[0m num-bigint v0.4.6
[1m[32m   Compiling[0m tracing-core v0.1.33
[1m[32m   Compiling[0m zerofrom-derive v0.1.5
[1m[32m   Compiling[0m yoke-derive v0.7.5
[1m[32m   Compiling[0m zerofrom v0.1.5
[1m[32m   Compiling[0m zerovec-derive v0.10.3
[1m[32m   Compiling[0m yoke v0.7.5
[1m[32m   Compiling[0m displaydoc v0.2.5
[1m[32m   Compiling[0m zerovec v0.10.4
[1m[32m   Compiling[0m serde_derive v1.0.217
[1m[32m   Compiling[0m tinystr v0.7.6
[1m[32m   Compiling[0m icu_locid v1.5.0
[1m[32m   Compiling[0m icu_provider_macros v1.5.0
[1m[32m   Compiling[0m icu_provider v1.5.0
[1m[32m   Compiling[0m icu_locid_transform v1.5.0
[1m[32m   Compiling[0m icu_collections v1.5.0
[1m[32m   Compiling[0m icu_properties v1.5.1
[1m[32m   Compiling[0m ref-cast-impl v1.0.23
[1m[32m   Compiling[0m futures-channel v0.3.31
[1m[32m   Compiling[0m tower-service v0.3.3
[1m[32m   Compiling[0m utf8parse v0.2.2
[1m[32m   Compiling[0m scopeguard v1.2.0
[1m[32m   Compiling[0m hyper v1.6.0
[1m[32m   Compiling[0m anstyle-parse v0.2.6
[1m[32m   Compiling[0m num-rational v0.4.2
[1m[32m   Compiling[0m icu_normalizer v1.5.0
[1m[32m   Compiling[0m idna_adapter v1.2.0
[1m[32m   Compiling[0m idna v1.0.3
[1m[32m   Compiling[0m tracing v0.1.41
[1m[32m   Compiling[0m getrandom v0.2.15
[1m[32m   Compiling[0m num-iter v0.1.45
[1m[32m   Compiling[0m num-complex v0.4.6
[1m[32m   Compiling[0m sync_wrapper v1.0.2
[1m[32m   Compiling[0m aho-corasick v1.1.3
[1m[32m   Compiling[0m zerocopy v0.7.35
[1m[32m   Compiling[0m regex-syntax v0.8.5
[1m[32m   Compiling[0m borrow-or-share v0.2.2
[1m[32m   Compiling[0m anstyle v1.0.10
[1m[32m   Compiling[0m bit-vec v0.8.0
[1m[32m   Compiling[0m anstyle-query v1.1.2
[1m[32m   Compiling[0m tower-layer v0.3.3
[1m[32m   Compiling[0m colorchoice v1.0.3
[1m[32m   Compiling[0m is_terminal_polyfill v1.70.1
[1m[32m   Compiling[0m anstream v0.6.18
[1m[32m   Compiling[0m tower v0.5.2
[1m[32m   Compiling[0m bit-set v0.8.0
[1m[32m   Compiling[0m fluent-uri v0.3.2
[1m[32m   Compiling[0m num v0.4.3
[1m[32m   Compiling[0m parking_lot v0.12.3
[1m[32m   Compiling[0m regex-automata v0.4.9
[1m[32m   Compiling[0m hyper-util v0.1.10
[1m[32m   Compiling[0m url v2.5.4
[1m[32m   Compiling[0m serde_urlencoded v0.7.1
[1m[32m   Compiling[0m http-body-util v0.1.2
[1m[32m   Compiling[0m ipnet v2.11.0
[1m[32m   Compiling[0m heck v0.5.0
[1m[32m   Compiling[0m vsimd v0.8.0
[1m[32m   Compiling[0m outref v0.5.2
[1m[32m   Compiling[0m lazy_static v1.5.0
[1m[32m   Compiling[0m log v0.4.25
[1m[32m   Compiling[0m uuid v1.13.1
[1m[32m   Compiling[0m strsim v0.11.1
[1m[32m   Compiling[0m clap_lex v0.7.4
[1m[32m   Compiling[0m base64 v0.22.1
[1m[32m   Compiling[0m mime v0.3.17
[1m[32m   Compiling[0m reqwest v0.12.12
[1m[32m   Compiling[0m clap_builder v4.5.27
[1m[32m   Compiling[0m uuid-simd v0.8.0
[1m[32m   Compiling[0m fraction v0.15.3
[1m[32m   Compiling[0m clap_derive v4.5.28
[1m[32m   Compiling[0m fancy-regex v0.14.0
[1m[32m   Compiling[0m referencing v0.29.0
[1m[32m   Compiling[0m email_address v0.2.9
[1m[32m   Compiling[0m num-cmp v0.1.0
[1m[32m   Compiling[0m bytecount v0.6.8
[1m[32m   Compiling[0m jsonschema v0.29.0
[1m[32m   Compiling[0m clap v4.5.28
[1m[32m   Compiling[0m jsonschema-cli v0.29.0 (/build/jsonschema-cli-0.29.0.tar.gz)
[1m[32m    Finished[0m `release` profile [optimized] target(s) in 1m 32s
Executing cargoInstallPostBuildHook
Finished cargoInstallPostBuildHook
Finished cargoBuildHook
buildPhase completed in 1 minutes 32 seconds
Running phase: checkPhase
@nix { "action": "setPhase", "phase": "checkPhase" }
Executing cargoCheckHook
cargoCheckHook flags: -j 4 --profile release --target x86_64-unknown-linux-gnu --offline -- --test-threads=4
[1m[32m   Compiling[0m syn v2.0.98
[1m[32m   Compiling[0m rustix v0.38.44
[1m[32m   Compiling[0m doc-comment v0.3.3
[1m[32m   Compiling[0m getrandom v0.3.1
[1m[32m   Compiling[0m predicates-core v1.0.9
[1m[32m   Compiling[0m difflib v0.4.0
[1m[32m   Compiling[0m termtree v0.5.1
[1m[32m   Compiling[0m assert_cmd v2.0.16
[1m[32m   Compiling[0m bitflags v2.8.0
[1m[32m   Compiling[0m linux-raw-sys v0.4.15
[1m[32m   Compiling[0m predicates v3.1.3
[1m[32m   Compiling[0m predicates-tree v1.0.12
[1m[32m   Compiling[0m bstr v1.11.3
[1m[32m   Compiling[0m console v0.15.10
[1m[32m   Compiling[0m synstructure v0.13.1
[1m[32m   Compiling[0m wait-timeout v0.2.1
[1m[32m   Compiling[0m fastrand v2.3.0
[1m[32m   Compiling[0m similar v2.7.0
[1m[32m   Compiling[0m linked-hash-map v0.5.6
[1m[32m   Compiling[0m tempfile v3.16.0
[1m[32m   Compiling[0m zerofrom-derive v0.1.5
[1m[32m   Compiling[0m yoke-derive v0.7.5
[1m[32m   Compiling[0m zerovec-derive v0.10.3
[1m[32m   Compiling[0m displaydoc v0.2.5
[1m[32m   Compiling[0m zerofrom v0.1.5
[1m[32m   Compiling[0m yoke v0.7.5
[1m[32m   Compiling[0m serde_derive v1.0.217
[1m[32m   Compiling[0m icu_provider_macros v1.5.0
[1m[32m   Compiling[0m ref-cast-impl v1.0.23
[1m[32m   Compiling[0m pin-project-internal v1.1.9
[1m[32m   Compiling[0m zerovec v0.10.4
[1m[32m   Compiling[0m ref-cast v1.0.23
[1m[32m   Compiling[0m clap_derive v4.5.28
[1m[32m   Compiling[0m tinystr v0.7.6
[1m[32m   Compiling[0m icu_locid v1.5.0
[1m[32m   Compiling[0m icu_collections v1.5.0
[1m[32m   Compiling[0m icu_provider v1.5.0
[1m[32m   Compiling[0m serde v1.0.217
[1m[32m   Compiling[0m icu_locid_transform v1.5.0
[1m[32m   Compiling[0m clap v4.5.28
[1m[32m   Compiling[0m pin-project v1.1.9
[1m[32m   Compiling[0m icu_properties v1.5.1
[1m[32m   Compiling[0m insta v1.42.1
[1m[32m   Compiling[0m icu_normalizer v1.5.0
[1m[32m   Compiling[0m serde_json v1.0.138
[1m[32m   Compiling[0m idna_adapter v1.2.0
[1m[32m   Compiling[0m idna v1.0.3
[1m[32m   Compiling[0m url v2.5.4
[1m[32m   Compiling[0m fluent-uri v0.3.2
[1m[32m   Compiling[0m serde_urlencoded v0.7.1
[1m[32m   Compiling[0m ahash v0.8.11
[1m[32m   Compiling[0m referencing v0.29.0
[1m[32m   Compiling[0m reqwest v0.12.12
[1m[32m   Compiling[0m email_address v0.2.9
[1m[32m   Compiling[0m jsonschema v0.29.0
[1m[32m   Compiling[0m jsonschema-cli v0.29.0 (/build/jsonschema-cli-0.29.0.tar.gz)
[1m[32m    Finished[0m `release` profile [optimized] target(s) in 58.81s
[1m[32m     Running[0m unittests src/main.rs (target/x86_64-unknown-linux-gnu/release/deps/jsonschema_cli-61df4afc79a9ae3e)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

[1m[32m     Running[0m tests/cli.rs (target/x86_64-unknown-linux-gnu/release/deps/cli-6bf5d206366139da)

running 6 tests
test test_no_instances ... ok
test test_invalid_schema ... ok
test test_invalid_instance ... ok
test test_multiple_instances ... ok
test test_version ... ok
test test_valid_instance ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s

Finished cargoCheckHook
checkPhase completed in 59 seconds
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing cargoInstallHook
Finished cargoInstallHook
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
shrinking RPATHs of ELF executables and libraries in /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0
shrinking /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0/bin/jsonschema-cli
checking for references to /build/ in /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0...
patching script interpreter paths in /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0
stripping (with command strip and flags -S -p) in  /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0/bin
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
Executing versionCheckPhase
Successfully managed to find version 0.29.0 in the output of the command /nix/store/j92lnyj16ybw7gsb7z5ya4v8ib3fadk5-jsonschema-cli-0.29.0/bin/jsonschema-cli --version
Version: 0.29.0
Finished versionCheckPhase
no Makefile or custom installCheckPhase, doing nothing

```

SuperSandro2000 replied:
But then we are already testing if it works, right? So we don't need our own tests, or what is the added benefit?

kachick replied:
It is completely up to their judgment which specs and behaviors should be tested. 

If package maintainers should always monitor and track the testing changes, it seems costly to me.

Thus, I hope adding some essential tests will reduce the reviewing tasks for the package maintainer.
However, I have much less experience in the role of a package maintainer. I will remove these package-side tests if they are unfit for our workflow. :bow:

kachick replied:
Removed in https://github.com/NixOS/nixpkgs/pull/382774/commits/5c1186349291858346a05457617da36b477ed002

SuperSandro2000 replied:
> If package maintainers should always monitor and track the testing changes, it seems costly to me.

But you would always follow upstream anyway... Don't take the packaging rules to serious 😅

kachick commented on pkgs/by-name/js/jsonschema-cli/package.nix:
```diff
@@ -0,0 +1,66 @@
+  };
+
+  meta = {
+    description = "A fast command-line tool for JSON Schema validation";
```
I have just extracted this description from their documents.

```console
> rg -F 'description = "' | wc -l
59712

> rg 'description = "[Aa] ' | wc -l
3061
```

Agreed that the absence of the 'A' prefix is the majority. Furthermore, a major tool fzf's description was renamed.

https://github.com/NixOS/nixpkgs/commit/bcd44e224fd68ce7d269b4f44d24c2220fd821e7#diff-1c964836b83557cecc31ba8ad2391b79db9deb010308dccc8ee86f099f2954eeR84

kachick replied:
https://github.com/NixOS/nixpkgs/blob/f47ba1d9e9218f3b8571331a297ae343bc07339f/pkgs/README.md?plain=1#L406

Sorry, overlooked 🙇‍♂️

kachick committed:
jsonschema-cli: remove leading indefinite article in description

https://github.com/NixOS/nixpkgs/blob/f47ba1d9e9218f3b8571331a297ae343bc07339f/pkgs/README.md?plain=1#L406

Co-authored-by: Tert0 <tert0byte@gmail.com>

kachick committed:
jsonschema-cli: remove unnecessary passthru.tests

Discussion: https://github.com/NixOS/nixpkgs/pull/382774/files#r1958260955

kachick commented:
Thank you Tert and Sandro for the polite review. 🙏
